File: Data Structures & Algorithms/climbing-stairs/submission-23.py
class Solution:
    def climbStairs(self, n: int) -> int:
        cache = [-1] * n

        def climb(i):
            if i == n:
                return 1
            if i > n:
                return 0
            
            if cache[i] != -1:
                return cache[i]
            
            cache[i] = climb(i + 1) + climb(i + 2)
            return cache[i]

        return climb(0)

File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-23.java
class Solution {
    public int climb(int i, int[] cost, int[] memo) {
        int n = cost.length;

        if (i >= cost.length) {
            return 0;
        }

        if (memo[i] != -1) {
            return memo[i];
        }
        memo[i] = cost[i] + Math.min(climb(i + 1, cost, memo), climb(i + 2, cost, memo));
        return memo[i];

    }
    public int minCostClimbingStairs(int[] cost) {
        int[] memo = new int[cost.length];
        for (int i = 0; i < cost.length; i++) {
            memo[i] = -1;
        }

        return Math.min(climb(0, cost, memo), climb(1, cost, memo));
    }
}
